C/decoder/decode.c

```diff
@@ -0,0 +1,200 @@
+	return (unsigned char)((*src >= 0 && *src <= 255) ? *src : 0);
+}
+
+
+static inline unsigned char *  create_byte_array(int size) {
+	return (unsigned char *)malloc(size * sizeof(unsigned char));
+}
+
+/*
+	decoder helper functions end
+*/
+
+
+/*
+	Base-83 decoder start
+*/
+
+float decode_to_int(const char * string, int start, int end) {
+	int value = 0, itr_1 = 0, itr_2 = 0;
+	for( itr_1 = start; itr_1 < end; itr_1 ++) {
+
+		int index = -1;
+		for(itr_2 = 0; itr_2 < 83; itr_2 ++) {
+			if (chars[itr_2] == string[itr_1]) {
+				index = itr_2;
+				break;
+			}
+		}
+
+		if (index == -1) return -1;
+		value = value * 83 + index;
+	}
+
+	return value;
+}
+
+/*
+	Base-83 decoder end
+*/
+
+/*
+	Blurhash functions start
+*/
+
+bool is_valid_blurhash(const char * string) {
+
+	const int hash_length = strlen(string);
+
+	if ( !string || strlen(string) < 6) return false;
+
+	int sizeFlag = decode_to_int(string, 0, 1);	//Get size from first character
+	int num_y = (int)floorf(sizeFlag / 9) + 1;
+	int num_x = (sizeFlag % 9) + 1;
+
+	if (hash_length != 4 + 2 * num_x * num_y) return false;
+
+	return true;
+}
+
+void decodeDc(int value, float * r, float * g, float * b) {
+	*r = sRGBToLinear(value >> 16); 	// R-component
+
+	*g = sRGBToLinear((value >> 8) & 255); // G-Component
+
+	*b = sRGBToLinear(value & 255);	// B-Component
+
+}
+
+
+void decodeAc(int value, float maximumValue, float * r, float * g, float * b) {
+
+    int quantR = (int)floorf(value / (19 * 19));
+	int quantG = (int)floorf(value / 19) % 19;
+	int quantB = (int)value % 19;
+
+
+	*r = signPow(((float)quantR - 9) / 9, 2.0) * maximumValue;
+
+	*g = signPow(((float)quantG - 9) / 9, 2.0) * maximumValue;
+
+	*b = signPow(((float)quantB - 9) / 9, 2.0) * maximumValue;
+}
+
+
+unsigned char * decode(const char * blurhash, int width, int height, int punch) {
+
+	if (! is_valid_blurhash(blurhash)) return NULL;
+
+	int size_flag = decode_to_int(blurhash, 0, 1);
+	int num_y = (int)floorf(size_flag / 9) + 1;
+	int num_x = (size_flag % 9) + 1;
+	int iter = 0;
+
+	float r = 0, g = 0, b = 0;
+
+	int quantizedMaxValue = decode_to_int(blurhash, 1, 2);
+	if (quantizedMaxValue == -1) return NULL;
+
+	float maxValue = ((float)(quantizedMaxValue + 1)) / 166;
+
+
+	int colors_size = num_x * num_y;
+	float colors[colors_size][3];
+
+	for(iter = 0; iter < colors_size; iter ++) {
+		if (iter == 0) {
+			int value = decode_to_int(blurhash, 2, 6);
+			if (value == -1) return NULL;
+			decodeDc(value, &r, &g, &b);
+			colors[iter][0] = r;
+			colors[iter][1] = g;
+			colors[iter][2] = b;
+
+		} else {
+			int value = decode_to_int(blurhash, 4 + iter * 2, 6 + iter * 2);
+			if (value == -1) return NULL;
+			decodeAc(value, maxValue * punch, &r, &g, &b);
+			colors[iter][0] = r;
+			colors[iter][1] = g;
+			colors[iter][2] = b;
+		}
+	}
+
+	int bytesPerRow = width * 4;
+	int x = 0, y = 0, i = 0, j = 0;
+	int intR = 0, intG = 0, intB = 0;
+
+	unsigned char * pixel_array = create_byte_array(bytesPerRow * height);
```

Comment:
prefer taking in a pre-allocated output parameter of pixels, rather than allocating them yourself. this makes it easier to bind to languages that have their own allocators.

Reply:
So, to address this I have added two different versions of the decode function, because the `decoder_main.c` is designed to work with old version, also since `decoder_main` CLI program is written in C, it can use `malloc` without any issues. Secondly, to make it portable with different programming languages with custom allocation methods, the second version of the function has been added, This accepts pointer to the pre-allocated array as an argument and manipulates it by reference.  The first version of the function just calls the second version internally, so we are still reusing the same code.

Please have a look at `decode` and `decodeToArray` function declarations in `decoder/decode.h`